### DISCUSSION TITLE
Tuya Zigbee thermostat

### DISCUSSION THREAD

**Discussion by gonzzzales** [arendst/Tasmota#11386] [Q&A] [answered]
Hello!
Arendst, thanks a lot for tasmota, it is really great!!!
I want to use Thermostat TS0601 with Zigbee2Tasmota, but have some problems, i dont undestand, how to change SetPoint temperature  through Tasmota. 
When i use ZbStatus3 0xD449, i can see it

`stat/tasmota_E35DD1/RESULT = {"ZbStatus3":[{"Device":"0xD449","Name":" termogolovka","IEEEAddr":"0x5C0272FFFEF1513A","ModelId":"TS0601","Manufacturer":"_TZE200_2dpplnsn","Endpoints":[1],"Config":["T01","~01.1"],"TempTarget":23,"Temperature":27.5,"Reachable":true,"LastSeen":132,"LastSeenEpoch":1616138831,"LinkQuality":196}]}`

My research has shown, that Tuya devices have their own zigbee cluster 0xEF00

`03.637 RSL: tele/tasmota_E35DD1/RESULT = {"ZbState":{"Status":33,"Device":"0xD449","Endpoint":"0x01","ProfileId":"0x0104","DeviceId":"0x0051","DeviceVersion":1,"InClusters":["0x0000","0x0004","0x0005","0xEF00"],"OutClusters":["0x0019","0x000A"]}}`

but how i can use it? I cant find any descriptions about it. Please, help me to unsderstand it. 
I tryed to use ZbSend { "device":"0xD449", "endpoint":"0x01", "ProfileId":"0x0104", "cluster":"0xEF00", "read":"0x????"} but unsuccessfully((

**Comment by s-hadinger** [comment, marked as answer]
Hi, Tuya type clusters (0xEF00 can't be read from. You need to wait for the device to report values.

To change values use: `ZbSend {"Device":"0xD449","Write":{"TuyaTempTarget":20.0}}`

Other possible attribute you can try are here (4th colum), although they may not all be supported nor useful for a TRV
```
  { Ztuya0,   CxEF00, 0x0070,  Z_(TuyaScheduleWorkdays), Cm1, 0 },
  { Ztuya0,   CxEF00, 0x0071,  Z_(TuyaScheduleHolidays), Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0101,  Z_(Power),                Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0102,  Z_(Power2),               Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0103,  Z_(Power3),               Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0104,  Z_(Power4),               Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0107,  Z_(TuyaChildLock),        Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0112,  Z_(TuyaWindowDetection),  Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0114,  Z_(TuyaValveDetection),   Cm1, 0 },
  { Ztuya1,   CxEF00, 0x0174,  Z_(TuyaAutoLock),         Cm1, 0 },
  { Zint16,   CxEF00, 0x0202,  Z_(TuyaTempTarget),       Cm_10, Z_MAPPING(Z_Data_Thermo, temperature_target) },
  { Zint16,   CxEF00, 0x0203,  Z_(LocalTemperature),     Cm_10, Z_MAPPING(Z_Data_Thermo, temperature) },  // will be overwritten by actual LocalTemperature
  { Ztuya2,   CxEF00, 0x0215,  Z_(TuyaBattery),          Cm1, 0 },   // TODO check equivalent?
  { Ztuya2,   CxEF00, 0x0266,  Z_(TuyaMinTemp),          Cm1, 0 },
  { Ztuya2,   CxEF00, 0x0267,  Z_(TuyaMaxTemp),          Cm1, 0 },
  { Ztuya2,   CxEF00, 0x0269,  Z_(TuyaBoostTime),        Cm1, 0 },
  { Ztuya2,   CxEF00, 0x026B,  Z_(TuyaComfortTemp),      Cm1, 0 },
  { Ztuya2,   CxEF00, 0x026C,  Z_(TuyaEcoTemp),          Cm1, 0 },
  { Zuint8,   CxEF00, 0x026D,  Z_(TuyaValvePosition),    Cm1, Z_MAPPING(Z_Data_Thermo, th_setpoint) },
  { Ztuya2,   CxEF00, 0x0272,  Z_(TuyaAwayTemp),         Cm1, 0 },
  { Ztuya2,   CxEF00, 0x0275,  Z_(TuyaAwayDays),         Cm1, 0 },
  { Ztuya4,   CxEF00, 0x0404,  Z_(TuyaPreset),           Cm1, 0 },
  { Ztuya4,   CxEF00, 0x0405,  Z_(TuyaFanMode),          Cm1, 0 },
  { Ztuya4,   CxEF00, 0x046A,  Z_(TuyaForceMode),        Cm1, 0 },
  { Ztuya4,   CxEF00, 0x046F,  Z_(TuyaWeekSelect),       Cm1, 0 },
```

**Reply by gonzzzales** [reply]
Thanks a lot. You solve my problem

**Reply by mundevx** [reply]
hi, i am working on tuya zigbee thermostat. 
            zb_zcl_custom_cluster_cmd_req_t cmd_req = {
                .address_mode = ZB_APS_ADDR_MODE_16_ENDP_PRESENT,
                .cluster_id = 0xEF00,
                .custom_cmd_id = ??,   //<---------
                .data.value = message->data, //10 bytes data
                .data.size = message->size,
                .direction = ZB_ZCL_CMD_DIRECTION_TO_CLI,
                .data.type = ??,   //<---------
                .profile_id = 0x0104,
                .zcl_basic_cmd.dst_addr_u.addr_short = 0,
                .zcl_basic_cmd.dst_endpoint = 1,
                .zcl_basic_cmd.src_endpoint = 1,   
                .dis_defalut_resp = 0,
                .manuf_code = message->info.header.manuf_code,
                .manuf_specific = 1,           
            };  
To set temperature on Tuya app, what are the values of custom_cmd_id & .data.type.
What should be data type for 10 bytes of frame. I used array, struct , etc. all types, but failed to make it working.
Can you please help me.

**Reply by s-hadinger** [reply]
I have no clue what you are trying to do. What commands did you try already ?

**Reply by mundevx** [reply]
0x01, 0x02, 0x10

**Reply by mundevx** [reply]
I am building cooling thermostat in tuya. i used all the thermostat cluster's attributes. My device is added in smartlife app. App can send datat and i receive that data. but when i send above data to tuya cluster 0xEF00, it is not working. But same command is working for data.size less than 9, i.e till 8. But, for setting thermostat temperature, i dont know the actual .data.type and custom_cmd_id to set temperature on Tuya app.

**Reply by mundevx** [reply]
Ok, let it clear simple.

server 0xEF00 Cluster Thermostat (Received callback) <--------------Send Request 7 Bytes---------------Tuya Coordintaor (Smart Life App)
                                                                           --------------------Respond 7 bytes---------------------> Success 

server 0xEF00 Cluster Thermostat (Received callback) <-----------------Send Request 10 Bytes-----------Tuya Coordintaor (Smart Life App)
                                                                          --------------------Respond 10 bytes-------------------> Failed

**Reply by s-hadinger** [reply]
Can you provide complete logs and the actual payloads? Count of bytes does not provide enough details.

**Reply by mundevx** [reply]
First is Thermostat State On-Off frame logs
THERMOSTAT On-OFF
cluster: 0xef00,  dst_endpoint: 1
command: 0x0,  size: 7
Frame: 0x3 0x57 0x1 0x1 0x0 0x1 0x0

Second is Thermostat temperature frame logs
THERMOSTAT TEMP
cluster: 0xef00,  dst_endpoint: 1
command: 0x0,  size: 10
Frame: 0x3 0x5d 0x32 0x2 0x0 0x4 0x0 0x0 0x0 0x13

**Reply by s-hadinger** [reply]
I still don't know what you have tried, but the structure of Tuya message is described here: https://tasmota.github.io/docs/Zigbee/#writing-zigbee-plugins

First line: `03570101000100`:
- `0357`: I don't remember what it is...
- `01`: dpid
- `01`: type = boolean
- `0001`: lenght in bytes (1)
- `00`: false (here Off)

Second is: `035d3202000400000013`
- `035d`: same as above
- `32`: dpid
- `02`: type = int32
- `0004`: length in bytes (4)
- `00000013`: value = 19

**Reply by mundevx** [reply]
These frames are what i receive, i have problem in sending back to update temperature value on smart life app. what will be tghe format then.

**Reply by mundevx** [reply]
I dont know what value is given to .data.type. 
For 7 bytes when send frame to coordinator, i know type_id is zigbee standard U56,
But for 10 bytes frame, there is tuya specific type_id. I dont know that type_id is that.

**Reply by s-hadinger** [reply]
Again I'm severely missing context. What I seemingly guess is that your question is unrelated to Tasmota. You are coding a thermostat In C that is supposed to connect to Smart Life App. 

I'm afraid you're in the wrong place. Tasmota is not related by any means to Tuya nor to Smart Life

**Reply by mundevx** [reply]
ok, Thank you
